src/symmetric_crypto/aes_256_gcm_pure/mod.rs

```diff
@@ -3,73 +3,102 @@
 
-    fn description() -> String {
-        format!(
-            "AES 256 GCM pure Rust (key bits: {}, nonce bits: {}, tag bits: {})",
-            KeyLength::to_usize() * 8,
-            NONCE_LENGTH * 8,
-            MAC_LENGTH * 8
-        )
-    }
+    type Key = Key<KEY_LENGTH>;
 
-    fn encrypt(
-        key: &Self::Key,
-        bytes: &[u8],
-        nonce: &Self::Nonce,
+    type Nonce = Nonce<NONCE_LENGTH>;
+
+    fn encrypt<R: RngCore + CryptoRng>(
+        rng: &mut R,
+        secret_key: &Self::Key,
+        plaintext: &[u8],
         additional_data: Option<&[u8]>,
     ) -> Result<Vec<u8>, CryptoCoreError> {
-        encrypt_combined(key, bytes, nonce, additional_data)
+        if plaintext.len() > MAX_PLAINTEXT_LENGTH {
+            return Err(CryptoCoreError::InvalidSize(format!(
+                "Plaintext is too large ({} bytes), max size: {} ",
+                plaintext.len(),
+                MAX_PLAINTEXT_LENGTH
+            )));
+        }
+        let nonce = Self::Nonce::new(rng);
+        // allocate correct byte number
+        let mut res: Vec<u8> = Vec::with_capacity(plaintext.len() + Self::ENCRYPTION_OVERHEAD);
+        res.extend_from_slice(nonce.as_bytes());
+        res.append(
+            &mut encrypt_combined(
+                secret_key.as_bytes(),
+                plaintext,
+                nonce.as_bytes(),
+                additional_data,
+            )
+            .map_err(|err| CryptoCoreError::EncryptionError(err.to_string()))?,
+        );
+        Ok(res)
     }
 
     fn decrypt(
-        key: &Self::Key,
-        bytes: &[u8],
-        nonce: &Self::Nonce,
+        secret_key: &Self::Key,
+        ciphertext: &[u8],
         additional_data: Option<&[u8]>,
     ) -> Result<Vec<u8>, CryptoCoreError> {
-        decrypt_combined(key, bytes, nonce, additional_data)
+        if ciphertext.len() < Self::ENCRYPTION_OVERHEAD {
+            return Err(CryptoCoreError::InvalidSize(format!(
+                "Ciphertext is too small ({} bytes), min size: {}",
+                ciphertext.len(),
+                Self::ENCRYPTION_OVERHEAD
+            )));
+        }
+        if ciphertext.len() > MAX_PLAINTEXT_LENGTH + Self::ENCRYPTION_OVERHEAD {
+            return Err(CryptoCoreError::InvalidSize(format!(
+                "Ciphertext is too large ({} bytes), max size: {} ",
+                ciphertext.len(),
+                MAX_PLAINTEXT_LENGTH + Self::ENCRYPTION_OVERHEAD
+            )));
+        }
+        // Read the nonce used for encryption. We know this will not fail since
+        // Self::Nonce::LENGTH < Self::ENCRYPTOION_OVERHEAD
+        let nonce = Self::Nonce::try_from(&ciphertext[..Self::Nonce::LENGTH])?;
+        decrypt_combined(
+            secret_key.as_bytes(),
+            &ciphertext[Self::Nonce::LENGTH..],
+            nonce.as_bytes(),
+            additional_data,
+        )
+        .map_err(|err| CryptoCoreError::EncryptionError(err.to_string()))
```

Comment:
Why mapping `CryptoCoreError` to `CryptoCoreError`?

Reply:
Indeed, I didn't see it